[ 테스트 ] cypress를 이용한 E2E 테스트 셋팅 및 목표 관련 E2E 테스트 코드 추가

MinaRoh opened the pull request:
close #346

모든 기능과 대부분의 오류가 이미 수정된 시기이기에 전체적으로 유저의 관점에서 서비스의 시작부터 끝까지 직접 사용해보며 발생할 오류를 찾는 것이 효율적이라 생각해 E2E 테스트를 도입하게 되었습니다. cypress가 프론트엔드에 최적화되어있으며 설정이 간단하고 테스트가 진행된 스냅샷 등을 확인 할 수 있어 선택했습니다.

## ✅ 작업 내용
- cypress 설치 및 셋팅
- 초기 로그인 관련 글로벌 테스트 코드 추가
- 목표 관련 테스트 코드 작성
  - 시나리오대로 작동하지 않는 오류 발견, 수정(https://github.com/FESI-4-4/slid-todo/commit/17811286d3206d4bb7e8608065150b22a67b7ca5, https://github.com/FESI-4-4/slid-todo/commit/3b38b6a073fb9ea394f0636bc0882588341a2268, https://github.com/FESI-4-4/slid-todo/pull/347/commits/309a4216347cfe136556ee056caa644f85b4c0c2, https://github.com/FESI-4-4/slid-todo/pull/347/commits/17d417930ee3729a3f7c3d62006ea967dc07af70)

24.11.20 추가
- 모바일, 태블릿, 데스크탑 환경에서 서로 다른 부분이 있어 goals 폴더 생성 후 `mobile.cy.ts`, `tablet.cy.ts,` `desktop.cy.ts` 파일을 만들어 따로 테스트를 진행했습니다.

### 사용법


`npm install`을 해줍니다.

프로젝트 루트에 `cypress.env.json` 파일을 만든 뒤 다음과 같이 테스트에 사용할 url과 테스트 로그인용 아이디 및 비밀번호를 적어 사용합니다.
(`cypress.env.json`은 `.gitignore`에 추가되어있습니다)

```json
{
  "CYPRESS_BASE_URL": "http://localhost:3000",
  "email": "이메일(test@test.com 등)",
  "password": "비밀번호"
}
```



cypress 설치(npm install 또는 앱 다운로드) -> npx cypress open -> E2E Testing -> Start E2E Testing in Chrome을 클릭하면 테스트용 브라우저가 오픈됩니다.
![image](https://github.com/user-attachments/assets/966f9b63-5893-4782-b9be-086789851f01)
테스트할 파일(현재는 goals.cy.ts)을 클릭하면 테스트가 시작되며 입력된 테스트 코드가 실행되는 모습을 볼 수 있습니다.

공통으로 자주 사용할 함수는 `cypress/support/commands.ts` 파일에 작성하고(`cypress/support/index.d.ts`도 적어주어야합니다) `cy.함수이름`으로 가져와 사용합니다.


그 외 자세한 테스트 코드는 `goals.cy.ts` 파일을 참고하여 작성해주시면 됩니다.(todos.cy.ts, notes.cy.ts, signup.cy.ts 등 핵심 기능들을 파일로 만들어 사용해주시면 될 것 같습니다)

+) it.only(), context.only() 처럼 only를 사용해 해당 테스트만 진행도 가능합니다.
+) viewport가 기본 1000,600 으로 고정이라 beforeEach에서 매번 화면 크기를 맞춰줍니다. 로그인시에도 화면 크기를 고정하고싶어 테스트 작성시에만 cypress.config.ts에서 고정해놓고 사용할 수 있도록 주석처리한 코드를 추가했습니다. 원하시는 디바이스 크기에 맞춰서 주석 해제하고 사용하시면 편리합니다.
+) 테스트 종료는 cypress 앱으로 돌아와 close를 누르면 됩니다.
![image](https://github.com/user-attachments/assets/01085eb3-0190-4aac-95d1-b56662f2681e)


## 📸 스크린샷 / GIF / Link

![Kapture 2024-11-19 at 14 44 52](https://github.com/user-attachments/assets/0b4e009b-1e53-40ea-89e5-4cb50d9753cb)


## 📌 이슈 사항

데스크탑 기준으로만 테스트가 진행되어 모바일과 태블릿 viewport로도 테스트를 추가로 진행해봐야 할 것 같습니다.
-> 완료했습니다. 

+) 시간 단축을 위해 cypress studio로 개발자가 직접 클릭하는 액션 등을 감지했다가 cypress 코드로 적어주는 기능을 사용해봐도 좋을 것 같습니다 https://inpa.tistory.com/entry/Cypress-Studio-%EB%8D%94%EC%9A%B1-%EA%B0%84%ED%8E%B8%ED%95%B4%EC%A7%84-%EC%9B%B9-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%9E%90%EB%8F%99%ED%99%94-%ED%88%B4

discussion에 cypress 관련 내용을 작성해두었습니다. #353

MinaRoh committed:
➕ #346 add: cypress 설치

MinaRoh committed:
🧪 #346 add: cypress config 추가

MinaRoh committed:
✅ #346 test: 글로벌 로그인 코드 추가

MinaRoh committed:
✅ #346 test: goals 테스트 코드 추가

MinaRoh committed:
🐛 #346 fix: 목표 상세 페이지에서 해당 목표가 사라지면 대시보드로 페이지 이동하도록 수정

MinaRoh committed:
🩹 #346 fix: 네비게이션 열기/닫기 버튼이 눌리면 aria-expanded 적용되도록 수정

MinaRoh committed:
🧪 #346 add: test context 추가 및 commands에 함수 추가

ahrrrl commented on cypress/e2e/goals.cy.ts:
```diff
@@ -0,0 +1,189 @@
+      });
+  });
+
+  context('목표 생성', () => {
```
오 테스트 코드로 체크하는 건 좋은데 목표 생성이 통과됐다는게 신기하군요
이거 네브에서 input에 목표 제목 적고 버튼 클릭하면 생성이 되나요? 엔터키를 누르면 등록은 되는데 제 기억으로 버튼을  마우스로 누르면 안된걸로 기억하는데

MinaRoh replied:
지금 확인해보니까 클릭으로는 생성이 안되네요... 엔터 입력이 click으로 지정되어있어서 그런 것 같습니다.. 버튼 눌러서 제출하는것도 따로 테스트 만들어서 수정하겠습니다~

MinaRoh committed:
✅ #346 test: goals에 버튼 클릭으로 할 일 추가 테스트 코드 추가

MinaRoh committed:
🐛 #346 fix: 목표 추가 버튼으로도 목표 추가되도록 수정

MinaRoh committed:
✅ #346 add: 목표 모바일, 태블릿 테스트 코드 추가

- goals 폴더를 만들어 각각 환경마다 cy.ts 파일을 만들었습니다.
- e2e.ts 파일에 nav 열기 닫기 함수를 추가했습니다.

MinaRoh committed:
🐛 #346 fix: 모바일, 태블릿 테스트 코드 오류 수정

- 모바일에서 목표 리스트가 너무 길어 최대 높이를 max-h-60으로 변경했습니다.
- 모바일 헤더의 햄버거 아이콘에 다른 환경에서와 마찬가지로 div로 감싸고 aria 요소를 추가했습니다.
- 모바일에서 네비게이션 닫기 버튼에 aria 요소를 추가했습니다.
- 목표 상세 페이지에서 해당 목표를 nav에서 삭제 후 대시보드로 이동할 때, 다시 nav를 확인하면 여전히 해당 목표가 남아있던 오류를 수정했습니다.

MinaRoh committed:
✅ #346 add: 모바일과 동일하게 삭제 테스트 수정

- 태블릿과 데스크탑에서도 모바일과 동일한 삭제 테스트를 하도록 수정했습니다.
- 그 외 오타를 수정했습니다.
